lib/webdriverio.js

```diff
@@ -170,28 +171,73 @@ let WebdriverIO = function (args, modifier) {
      * check if the last and current promise got rejected. If so we can throw the error.
      */
     let reject = function (err, onRejected) {
-        if (!options.isWDIO && typeof onRejected === 'function') {
+        if (!options.isWDIO && !options.screenshotOnReject && typeof onRejected === 'function') {
             delete err.screenshot
             return onRejected(err)
-        } else if (typeof onRejected === 'function') {
-            onRejected(err)
+        }
+
+        if (typeof onRejected !== 'function') {
+            onRejected = () => {}
         }
 
         if (!this) {
+            onRejected(err)
             throw err
-        } else if (this.depth !== 0) {
+        }
+
+        if (this.depth !== 0) {
+            onRejected(err)
             return this.promise
         }
 
-        /**
-         * take screenshot only if screenshotPath is given
-         */
-        if (typeof options.screenshotPath !== 'string' || err.shotTaken) {
+        const shouldTakeScreenshot = options.screenshotOnReject || typeof options.screenshotPath === 'string'
+
+        if (!shouldTakeScreenshot || err.shotTaken || insideCommand('screenshot', this)) {
+            onRejected(err)
             return throwException(err, stacktrace)
         }
 
-        let screenshotPath = path.isAbsolute(options.screenshotPath)
-            ? options.screenshotPath : path.join(process.cwd(), options.screenshotPath)
+        return takeScreenshot.call(this, err)
+            .catch((e) => logger.log('\tFailed to take screenshot on reject:', e))
```

Comment:
> I would prefer to have that catch callback as 2nd argument to the then function

why? In current format code looks more readable

> Also let's do logger.error

we already handling an error. And in most cases the cause of screenshot fail will be the same as the original error